lib/Quantization/Quantization.cpp

```diff
@@ -111,333 +192,146 @@ static bool canBeQuantized(const Node *node) {
+  for (unsigned i = 0, e = node.getNumInputs(); i < e; ++i) {
+    if (node.getNthInput(i).getElementType() != ElemKind::FloatTy) {
       return false;
     }
   }
 
   return true;
 }
 
-/// Quantize the \p node such that all floating point inputs and outputs
-/// are quantized to int8 type with some scale and offset.
-/// \returns Quantized node.
-///
-/// \param F Function which holds the non quantized \p node.
-/// \param node Node to be quantized.
-/// \param quantizedInputs Array of already quantized inputs to the result node.
-/// \param qParams Tensor quantization parameters for all outputs of the
-///        \p node.
-static Node *quantizeNode(Function *F, Node *node,
-                          llvm::MutableArrayRef<NodeValue> quantizedInputs,
-                          llvm::ArrayRef<TensorQuantizationParams> qParams) {
-  Node *quantizedNode{};
-
-  switch (node->getKind()) {
-  case Kinded::Kind::FullyConnectedNodeKind: {
-    auto *FC = cast<FullyConnectedNode>(node);
-    assert(quantizedInputs.size() == 3 && "Invalid number of inputs");
-    assert(qParams.size() == 1 && "Invalid number of quantized outputs");
-    auto QT =
-        F->getParent()->uniqueType(ElemKind::Int8QTy, FC->getResult().dims(),
-                                   qParams[0].scale, qParams[0].offset);
-    quantizedNode =
-        F->createFullyConnected(FC->getName(), quantizedInputs[0],
-                                quantizedInputs[1], quantizedInputs[2], QT);
-    break;
-  }
+Node &FunctionQuantizer::morphNode(Node &node) {
+  // FIXME: Right now, the TensorQuantizationParams only tracks one
+  // type per NodeValue, whereas we would need one type for the output and
+  // one for user of that value. E.g., for
+  // val = node
+  // = use1 val
+  // = use2 val
+  //
+
+  // We would want to track independently the types for the result of node, the
+  // use of val in use1, and the use of val in use2, in respectively outTy,
+  // inTy1, and inTy2, so that we can generate: outTy = node inTy1 = cast outTy
+  // = use1 inTy1
+  // inTy2 = cast outTy
+  // = use2 inTy2
+  //
+  //
+  // But instead what we track only one type like this:
+  // outTy = node
+  // = use1 outTy
+  // = use2 outTy
+  //
+  // However, sometimes outTy is not suitable for the input (we fix those in
+  // postProcessing) and sometimes, the outTy is not suitable for the output
+  // itself!
+  // What this means basically is outTy encodes the inTy constraints whereas
+  // they may disagree.
+  // The following switch fixes outTy for the few nodes where inTy and outTy
+  // can disagree.
+  // This happens for cases where for instance, the quantized parameters for the
+  // output have a different scale than the input, whereas the operation itself
+  // doesn't allow that.
+  //
+  // E.g., the constraints for `outTy = op inTy` are `outTy == inTy`, but the
+  // quantization profiling gave different types to outTy and inTy.
+  switch (node.getKind()) {
+  default:
```

Comment:
Nit: `default` at the beginning feels kind of weird, since I think it's at the end of the switch everywhere else.

Reply:
Sure!

Reply:
Done